[c10d][nvshmem] add nvshmem build rules and dependency for libtorch_cuda

jushg opened the pull request:
Summary:
Add guarded build option for nvshmem-related c10d code with `-c fbcode.caffe2_use_nvshmem`

Guarded clause include nvshmem device + host code (static-linked) + these 2 files:
- `torch/csrc/distributed/c10d/symm_mem/NVSHMEMSymmetricMemory.cu`
-    `torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu`




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

iremyux commented:
Adding `ciflow/win-arm64` label to trigger Windows Arm64 CI and its test purposes - nothing about this PR specifically. (It should not effect the acceptance of the PR even if it fails.)

jushg committed:
[c10d][nvshmem] add nvshmem build rules and dependency for libtorch_cuda (#159562)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/159562

Add guarded build option for nvshmem-related c10d code with `-c fbcode.caffe2_use_nvshmem`

Guarded clause include nvshmem device + host code (static-linked) + these 2 files:
- `torch/csrc/distributed/c10d/symm_mem/NVSHMEMSymmetricMemory.cu`
-    `torch/csrc/distributed/c10d/symm_mem/nvshmem_extension.cu`

internals:
For nvshmem we need to build the torch src code and libnvshmem with separate compilation, then dlink them together, hence the complicated src code build rule.

Reviewed By: kwen2501

Differential Revision: D78813568
